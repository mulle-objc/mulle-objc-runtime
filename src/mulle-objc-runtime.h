//
//  mulle_objc.h
//  mulle-objc-runtime
//
//  Created by Nat! on 10.07.16.
//  Copyright (c) 2016 Nat! - Mulle kybernetiK.
//  Copyright (c) 2016 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_objc_runtime_h__
#define mulle_objc_runtime_h__

/*
 * If you want to include <mulle-objc-runtime/mulle-objc-runtime.h> you
 * should be compiling with an Objective-C compiler, that defines the
 * following defines properly. Do not use .c  use .m with mulle-objc-runtime.h
 */

// catch this early
#if ! defined( __MULLE_OBJC_TPS__) && ! defined( __MULLE_OBJC_NO_TPS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_TPS__ or __MULLE_OBJC_NO_TPS__)"
#endif
#if defined( __MULLE_OBJC_TPS__) && defined( __MULLE_OBJC_NO_TPS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_TPS__ and __MULLE_OBJC_NO_TPS__)"
#endif
#if ! defined( __MULLE_OBJC_FCS__) && ! defined( __MULLE_OBJC_NO_FCS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_FCS__ or __MULLE_OBJC_NO_FCS__)"
#endif
#if defined( __MULLE_OBJC_FCS__) && defined( __MULLE_OBJC_NO_FCS__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_FCS__ and __MULLE_OBJC_NO_FCS__)"
#endif
#if ! defined( __MULLE_OBJC_TAO__) && ! defined( __MULLE_OBJC_NO_TAO__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (or define either __MULLE_OBJC_TAO__ or __MULLE_OBJC_NO_TAO__)"
#endif
#if defined( __MULLE_OBJC_TAO__) && defined( __MULLE_OBJC_NO_TAO__)
# error "Use the mulle-clang 17.0.6.0 (or newer) compiler to compile mulle-objc code (do not define both __MULLE_OBJC_TAO__ and __MULLE_OBJC_NO_TAO__)"
#endif

//
// these should have been defined by the compiler already, but for
// mulle-objc-runtime testing with gcc its convenient to add these if missing.
//
#ifndef __MULLE_OBJC_UNIVERSEID__
# define __MULLE_OBJC_UNIVERSEID__  0
#endif
#ifndef __MULLE_OBJC_UNIVERSENAME__
# define __MULLE_OBJC_UNIVERSENAME__  NULL
#endif

#if defined( __MULLE_OBJC_TPS__) && __MULLE_OBJC_UNIVERSEID__ != 0
# error "TPS can only be used in the default universe not in " #__MULLE_OBJC_UNIVERSENAME__
#endif


#include "include.h"

#include "mulle-metaabi.h"

#include "mulle-objc-atomicpointer.h"
#include "mulle-objc-builtin.h"
#include "mulle-objc-call.h"
#include "mulle-objc-class.h"
#include "mulle-objc-classpair.h"
#include "mulle-objc-class-convenience.h"
#include "mulle-objc-class-initialize.h"
#include "mulle-objc-class-lookup.h"
#include "mulle-objc-class-search.h"
#include "mulle-objc-class-struct.h"
#include "mulle-objc-csvdump.h"
#include "mulle-objc-fastclasstable.h"
#include "mulle-objc-fastmethodtable.h"
#include "mulle-objc-infraclass.h"
#include "mulle-objc-ivar.h"
#include "mulle-objc-ivarlist.h"
#include "mulle-objc-kvccache.h"
#include "mulle-objc-load.h"
#include "mulle-objc-metaclass.h"
#include "mulle-objc-method.h"
#include "mulle-objc-methodidconstants.h"
#include "mulle-objc-methodlist.h"
#include "mulle-objc-object.h"
#include "mulle-objc-objectheader.h"
#include "mulle-objc-object-convenience.h"
#include "mulle-objc-property.h"
#include "mulle-objc-propertylist.h"
#include "mulle-objc-retain-release.h"
#include "mulle-objc-signature.h"
#include "mulle-objc-super.h"
#include "mulle-objc-symbolizer.h"
#include "mulle-objc-taggedpointer.h"
#include "mulle-objc-try-catch-finally.h"
#include "mulle-objc-universe.h"
#include "mulle-objc-universe-class.h"
#include "mulle-objc-universe-exception.h"
#include "mulle-objc-universe-fail.h"
#include "mulle-objc-universe-global.h"
#include "mulle-objc-universe-struct.h"
#include "mulle-objc-uniqueid.h"
#include "mulle-objc-uniqueidarray.h"
#include "mulle-objc-version.h"
#include "mulle-objc-walktypes.h"



// add some functions to mulle-vararg for ObjC

#define mulle_vararg_count_ids( args, obj) \
   mulle_vararg_count_pointers( args, obj)
#define mulle_vararg_count_objects( args, obj) \
   mulle_vararg_count_pointers( args, obj)

#define mulle_vararg_next_id( args) \
   mulle_vararg_next_pointer( args, id)
#define mulle_vararg_next_object( args, type) \
   mulle_vararg_next_pointer( args, type)

#define mulle_vararg_push_id( args, value) \
   _mulle_vararg_push( args, id, value)
#define mulle_vararg_push_object( args, value) \
   mulle_vararg_push_id( args, value)

//
// add this before return of int main( argc, char **argc) when running
// named universes and leak checking
//
static inline void  mulle_objc_global_finish( void)
{
   struct _mulle_objc_universe   *universe;

   universe = mulle_objc_global_get_universe_inline( __MULLE_OBJC_UNIVERSEID__);
   if( universe && ! _mulle_objc_universe_is_default( universe) && universe->config.pedantic_exit)
      _mulle_objc_universe_release( universe);
}



#ifdef __has_include
# if __has_include( "_mulle-objc-runtime-versioncheck.h")
#  include "_mulle-objc-runtime-versioncheck.h"
# endif
#endif


#include <assert.h>
#include <stdarg.h>
#include <setjmp.h>

#endif

