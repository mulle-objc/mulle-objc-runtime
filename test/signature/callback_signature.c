//
//  main.c
//  test-runtime-2
//
//  Created by Nat! on 19/11/14.
//  Copyright (c) 2014 Mulle kybernetiK. All rights reserved.
//
#ifndef __MULLE_OBJC__
# define __MULLE_OBJC_NO_TPS__
# define __MULLE_OBJC_FCS__
# if defined( DEBUG) || ! defined( __OPTIMIZE__)
#  define __MULLE_OBJC_TAO__
# else
#  define __MULLE_OBJC_NO_TAO__
# endif
#endif



#include <mulle-objc-runtime/mulle-objc-runtime.h>

#include <stdio.h>


static char  *signatures[] =
{
   "{anonymous={?=c*i}{?=c@}[2{?=@*}](?=cq)}",
   "{?=ccfc{?=fc}}",
   "{?=c^{?=f^*}}",
   "*",
   "Q",
   "^v",
   "^s",
   "@\"NSData\"",
   0
};



struct callback_info
{
   FILE                  *fp;
   ptrdiff_t             invocation_offset;
   unsigned int          index;
   struct mulle_buffer   stack;
};


static void   callback( char *type, struct mulle_objc_typeinfo *info, void *userinfo)
{
   struct callback_info  *p = userinfo;
   FILE                  *fp;
   int                   state;

   if( ! info)
   {
      fprintf( fp, "\tNULL\n");
      return;
   }

   switch( *type)
   {
   case _C_STRUCT_B : mulle_buffer_add_byte( &p->stack, *type); break;
   case _C_STRUCT_E : mulle_buffer_pop_byte( &p->stack);        break;
   case _C_UNION_B  : mulle_buffer_add_byte( &p->stack, *type); break;
   case _C_UNION_E  : mulle_buffer_pop_byte( &p->stack);        break;
   case _C_ARY_B    : mulle_buffer_add_byte( &p->stack, *type); break;
   case _C_ARY_E    : mulle_buffer_pop_byte( &p->stack);        break;
   }

   state = mulle_buffer_get_last_byte( &p->stack);

   if( state != _C_STRUCT_E && state != _C_UNION_E && state != _C_ARY_E)
   {
      info->invocation_offset = (int32_t) mulle_address_align( p->invocation_offset,
                                                               p->index == 0
                                                               ? alignof( long double)
                                                               : info->bits_struct_alignment / 8);
      p->invocation_offset = info->invocation_offset + info->natural_size;
      p->index++;
   }


   fprintf( p->fp, "\t(p) type              = %s\n",  type);
   fprintf( p->fp, "\tnatural_size          = %zd\n", info->natural_size);
   fprintf( p->fp, "\tnatural_alignment     = %d\n",  (int) info->natural_alignment);
   fprintf( p->fp, "\tmember_type_start     = %s\n",  info->member_type_start ? info->member_type_start : "NULL");
   fprintf( p->fp, "\thas_object            = %d\n",  info->has_object);
   fprintf( p->fp, "\thas_retainable_type   = %d\n",  info->has_retainable_type);
   fprintf( p->fp, "\tn_members             = %ld\n", (long) info->n_members);
 // fprintf( p->fp, "\tinvocation_offset     = %td\n", (ptrdiff_t) info->invocation_offset);
   fprintf( p->fp, "\tbits_size             = %zd\n", info->bits_size);
   fprintf( p->fp, "\tbits_struct_alignment = %d\n\n",  (int) info->bits_struct_alignment);
}



static void   test( char *signature)
{
   unsigned int                 i, n;
   char                         *next;
   struct mulle_objc_typeinfo   type_info;
   struct callback_info         callback_info;
   char                         *type;

   printf( "type=\"%s\":\n", signature);

   memset( &type_info, 0xDEAD, sizeof( type_info)); // just to mess things up :)

   callback_info = (struct callback_info){ .fp = stdout,
                                           .invocation_offset = 0,
                                           .stack = MULLE_BUFFER_DATA( NULL)
                                          };
   type = _mulle_objc_type_parse( signature,
                                  0,
                                  &type_info,
                                  _mulle_objc_signature_supply_scalar_typeinfo,
                                  callback,
                                  &callback_info);
   mulle_buffer_done( &callback_info.stack);

   if( ! type)
   {
      printf( "SYNTAX ERROR IN SIGNATURE: \"%s\"\n", signature);
      return;
   }
   printf( "\n");
}




int   main( int argc, const char * argv[])
{
   char  **signature;

   for( signature = signatures; *signature; signature++)
   {
      test( *signature);
   }

   return 0;
}


