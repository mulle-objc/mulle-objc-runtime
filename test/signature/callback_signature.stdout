type="{anonymous={?=c*i}{?=c@}[2{?=@*}](?=cq)}":
	(p) type              = {anonymous={?=c*i}{?=c@}[2{?=@*}](?=cq)}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = {?=c*i}{?=c@}[2{?=@*}](?=cq)}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = {?=c*i}{?=c@}[2{?=@*}](?=cq)}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = c*i}{?=c@}[2{?=@*}](?=cq)}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = c*i}{?=c@}[2{?=@*}](?=cq)}
	natural_size          = 1
	natural_alignment     = 1
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 8
	bits_struct_alignment = 8

	(p) type              = *i}{?=c@}[2{?=@*}](?=cq)}
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 1
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64

	(p) type              = i}{?=c@}[2{?=@*}](?=cq)}
	natural_size          = 4
	natural_alignment     = 4
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 32
	bits_struct_alignment = 32

	(p) type              = }{?=c@}[2{?=@*}](?=cq)}
	natural_size          = 24
	natural_alignment     = 8
	member_type_start     = c*i}{?=c@}[2{?=@*}](?=cq)}
	has_object            = 0
	has_retainable_type   = 1
	n_members             = 3
	bits_size             = 192
	bits_struct_alignment = 64

	(p) type              = {?=c@}[2{?=@*}](?=cq)}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = c@}[2{?=@*}](?=cq)}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = c@}[2{?=@*}](?=cq)}
	natural_size          = 1
	natural_alignment     = 1
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 8
	bits_struct_alignment = 8

	(p) type              = @}[2{?=@*}](?=cq)}
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 1
	has_retainable_type   = 1
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64

	(p) type              = }[2{?=@*}](?=cq)}
	natural_size          = 16
	natural_alignment     = 8
	member_type_start     = c@}[2{?=@*}](?=cq)}
	has_object            = 1
	has_retainable_type   = 1
	n_members             = 2
	bits_size             = 128
	bits_struct_alignment = 64

	(p) type              = [2{?=@*}](?=cq)}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = {?=@*}](?=cq)}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 2
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = {?=@*}](?=cq)}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = @*}](?=cq)}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = @*}](?=cq)}
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 1
	has_retainable_type   = 1
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64

	(p) type              = *}](?=cq)}
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 1
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64

	(p) type              = }](?=cq)}
	natural_size          = 16
	natural_alignment     = 8
	member_type_start     = @*}](?=cq)}
	has_object            = 1
	has_retainable_type   = 1
	n_members             = 2
	bits_size             = 128
	bits_struct_alignment = 64

	(p) type              = ](?=cq)}
	natural_size          = 32
	natural_alignment     = 8
	member_type_start     = {?=@*}](?=cq)}
	has_object            = 1
	has_retainable_type   = 1
	n_members             = 2
	bits_size             = 256
	bits_struct_alignment = 64

	(p) type              = (?=cq)}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = cq)}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = cq)}
	natural_size          = 1
	natural_alignment     = 1
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 8
	bits_struct_alignment = 8

	(p) type              = q)}
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64

	(p) type              = )}
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = cq)}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 2
	bits_size             = 64
	bits_struct_alignment = 64

	(p) type              = }
	natural_size          = 80
	natural_alignment     = 8
	member_type_start     = {?=c*i}{?=c@}[2{?=@*}](?=cq)}
	has_object            = 1
	has_retainable_type   = 1
	n_members             = 4
	bits_size             = 640
	bits_struct_alignment = 64


type="{?=ccfc{?=fc}}":
	(p) type              = {?=ccfc{?=fc}}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = ccfc{?=fc}}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = ccfc{?=fc}}
	natural_size          = 1
	natural_alignment     = 1
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 8
	bits_struct_alignment = 8

	(p) type              = cfc{?=fc}}
	natural_size          = 1
	natural_alignment     = 1
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 8
	bits_struct_alignment = 8

	(p) type              = fc{?=fc}}
	natural_size          = 4
	natural_alignment     = 4
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 32
	bits_struct_alignment = 32

	(p) type              = c{?=fc}}
	natural_size          = 1
	natural_alignment     = 1
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 8
	bits_struct_alignment = 8

	(p) type              = {?=fc}}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = fc}}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = fc}}
	natural_size          = 4
	natural_alignment     = 4
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 32
	bits_struct_alignment = 32

	(p) type              = c}}
	natural_size          = 1
	natural_alignment     = 1
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 8
	bits_struct_alignment = 8

	(p) type              = }}
	natural_size          = 8
	natural_alignment     = 4
	member_type_start     = fc}}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 2
	bits_size             = 64
	bits_struct_alignment = 32

	(p) type              = }
	natural_size          = 20
	natural_alignment     = 4
	member_type_start     = ccfc{?=fc}}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 5
	bits_size             = 160
	bits_struct_alignment = 32


type="{?=c^{?=f^*}}":
	(p) type              = {?=c^{?=f^*}}
	natural_size          = 0
	natural_alignment     = 0
	member_type_start     = c^{?=f^*}}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 0
	bits_struct_alignment = 0

	(p) type              = c^{?=f^*}}
	natural_size          = 1
	natural_alignment     = 1
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 8
	bits_struct_alignment = 8

	(p) type              = ^{?=f^*}}
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64

	(p) type              = }
	natural_size          = 16
	natural_alignment     = 8
	member_type_start     = c^{?=f^*}}
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 2
	bits_size             = 128
	bits_struct_alignment = 64


type="*":
	(p) type              = *
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 1
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64


type="Q":
	(p) type              = Q
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64


type="^v":
	(p) type              = ^v
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64


type="^s":
	(p) type              = ^s
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 0
	has_retainable_type   = 0
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64


type="@"NSData"":
	(p) type              = @"NSData"
	natural_size          = 8
	natural_alignment     = 8
	member_type_start     = NULL
	has_object            = 1
	has_retainable_type   = 1
	n_members             = 0
	bits_size             = 64
	bits_struct_alignment = 64


